### PR TITLE
Initial autonomous code

### PR DESCRIPTION
**Pull request by mlists**
The current intention is to use the RamseteController built into wpilib for our trajectory tracking. Trajectories are currently generated manually but tools like pathweaver are potentially usable as well.

### CODE REVIEW

**Comment by mlists** [comment]
Currently the controller is able to track positions correctly but does not seem to obey the acceleration or velocity constraints that we specify.

File: controllers/shooter.py [hunk 1]
```diff
@@ -81,12 +97,13 @@ def searching(self) -> None:
...
         else:
             # TODO test this
-            # pose: wpilib.geometry.Pose2d = self.chassis.get_pose()
-            # rel: wpilib.geometry.Pose2d = self.TARGET_POSITION.relativeTo(pose)
-            # rel_heading = rel.rotation().radians()
-            # self.turret.scan(rel_heading)
-
-            self.turret.scan(0)  # TODO remove this
+            pose: wpilib.geometry.Pose2d = self.chassis.get_pose()
+            rel: wpilib.geometry.Pose2d = pose - self.TARGET_POSITION
```
Comment (auscompgeek) [inline controllers/shooter.py]: I think this type hint is incorrect - the difference between two Pose2ds should be a Transform2d (since you would add a Pose2d and Transform2d to get a new Pose2d).

File: controllers/shooter.py [hunk 2]
```diff
@@ -81,12 +97,13 @@ def searching(self) -> None:
...
+            rel_heading = (
+                math.atan2(rel.translation().Y(), rel.translation().X())
+                + pose.rotation().radians()
+            )
```
Comment (auscompgeek) [inline controllers/shooter.py]: It seems like this should be constrained to [-pi,pi], but it might be better to do it like this:

```python
translation = rel.translation()
rel_heading = Rotation2d(translation.y, translation.x) + pose.rotation()

File: autonomous/shoot_move_shoot.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        )
+
+
+class _3Right3(ShootMoveShootBase):
+    MODE_NAME = "3RIGHT3"
+
+    def setup(self):
+        super().setup()
+        self.start_pose = to_pose(3.459, -0.705, 0)
+        self.chassis.reset_odometry(self.start_pose)
```
Comment (auscompgeek) [inline autonomous/shoot_move_shoot.py]: This is probably something you want to do in `on_enable` rather than in setup, otherwise all your autonomous modes will clobber each other's starting pose.

File: autonomous/shoot_move_shoot.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        self.end_pose = to_pose(8.163, -0.705, 0)
+        self.waypoints = [
+            geometry.Translation2d(7.077, -0.705),
+            geometry.Translation2d(7.992, -0.705),
+        ]
+        self.trajectory_config = trajectory.TrajectoryConfig(
+            maxVelocity=1.5, maxAcceleration=1
+        )
```
Comment (auscompgeek) [inline autonomous/shoot_move_shoot.py]: You seem to be redefining `trajectory_config` here. This clobbers the `setKinematics` call in the super `setup`.

If you move this to `__init__`, you could move the `self.path` to the super `setup` as well.

File: autonomous/shoot_move_shoot.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        self.controller.setTolerance(tolerance)
+        self.trajectory_config = trajectory.TrajectoryConfig(
+            maxVelocity=1, maxAcceleration=1
+        )
+        self.gen = trajectory.TrajectoryGenerator()
+
+    def setup(self):
+        self.trajectory_config.setKinematics(self.chassis.kinematics)
+        self.trajectory_num = 0
+        self.trajectory_max = 1
```
Comment (auscompgeek) [inline autonomous/shoot_move_shoot.py]: Given we override `__init__` here, we should probably move things that don't depend on the injected variables to it.

**Review by auscompgeek — DISMISSED** [review]
Seems reasonable for a first iteration. A few comments/bugs I've noticed.

Please rebase - the changes to the chassis have been merged separately now.

File: components/chassis.py [hunk 6]
```diff
@@ -130,3 +133,11 @@ def reset_odometry(self, pose: Pose2d) -> None:
...
         self.right_encoder.setPosition(0)
         self.odometry.resetPosition(pose, self._get_heading())
+
+    def angle_to_target(self) -> float:
+        """Return the theoretical angle (in radians) to the target based on odometry"""
+        pose: Pose2d = self.chassis.get_pose()
+        rel: Transform2d = pose - self.TARGET_POSITION
+        translation = rel.translation()
+        rel_heading = Rotation2d(translation.y, translation.x) + pose.rotation()
+        return rel_heading.radians()
```
Comment (auscompgeek) [inline components/chassis.py]: Now that this has been factored out, I can see what this is actually trying to do… and I'm not sure that the maths is correct.

I think it would be prudent to have a method to find the heading to _any_ position (as a Translation2d) on the field. You would then just subtract the two translations and turn that into a Rotation2d.

We could then have a convenience method `get_angle_to_power_port` (let's use the game's terms) which calls `get_angle_to` or whatever we call it with the correct coordinates.

**Review by OliverW10 — APPROVED** [review]

File: autonomous/shoot_move_shoot.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+from wpilib.trajectory import constraint
+from magicbot import AutonomousStateMachine, state
+
+from components.chassis import Chassis
+from components.indexer import Indexer
+from components.shooter import Shooter
+from controllers.shooter import ShooterController
+
+
+def to_pose(x: float, y: float, heading: float) -> geometry.Pose2d:
```
Comment (auscompgeek) [inline autonomous/shoot_move_shoot.py:15]: We should replace all calls to this to just call the Pose2d constructor, now that there is an overload that takes the angle directly rather than a Rotation2d object.

File: autonomous/shoot_move_shoot.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    def setup(self):
+        self.trajectory_config.setKinematics(self.chassis.kinematics)
+        self.trajectory_config.addConstraint(
+            constraint.DifferentialDriveVoltageConstraint(
+                self.chassis.ff_calculator, self.chassis.kinematics, 10
+            )
+        )
+        self.end_ranges = []
+        self.paths = []
+        for i in range(self.trajectory_max):
```
Comment (auscompgeek) [inline autonomous/shoot_move_shoot.py:55]: We should probably refactor this at some point so it's impossible to get all these out of sync.

File: components/indexer.py [hunk 9]
```diff
@@ -15,7 +15,7 @@ class Indexer:
     intake_left_motor: wpilib.interfaces.SpeedController  # Looking from behind the robot
     intake_right_motor: wpilib.interfaces.SpeedController  # Looking from behind the robot
 
-    indexer_speed = tunable(0.6)
+    indexer_speed = tunable(0.4)
```
Comment (auscompgeek) [inline components/indexer.py:18]: Was this intentional? What's the reason behind this change?

File: controllers/shooter.py [hunk 10]
```diff
@@ -151,17 +163,24 @@ def fire_input(self) -> None:
...
+    def aimed_at_target(self) -> bool:
+        target_data = self.target_estimator.get_data()
+        return abs(target_data.angle) < self.find_allowable_angle()
+
+    def find_allowable_angle(self) -> float:
         """
         Find the maximum angle by which the turret can be misaligned to still score a hit
         Currently does not consider angle from target
-        dist: planar distance from the target
+        dist: the maximum value of our lookup table
```
Comment (auscompgeek) [inline controllers/shooter.py:181]: Does this still belong in the docstring?

File: components/indexer.py (hunk 9 above)
Comment (mlists) [reply components/indexer.py]: I don't think this was intentional, will revert
